### PR TITLE
[Issues/372] Bad code in schema registry client

### LINKED ISSUE #372: Bad code in CachedSchemaRegistryClient
Description
===========
`compatibility` doesn't exist, so this could cause a exception, is that the purpose?
https://github.com/confluentinc/confluent-kafka-python/blob/71217575e65d5652508d80d0532314acd190fd85/confluent_kafka/avro/cached_schema_registry_client.py#L314

This doesn't `return`, seems to be a bug, so the function would actually return None.
https://github.com/confluentinc/confluent-kafka-python/blob/71217575e65d5652508d80d0532314acd190fd85/confluent_kafka/avro/cached_schema_registry_client.py#L271

What's the purpose of this exception catch? if you are going to raise it without logging or don't care?
https://github.com/confluentinc/confluent-kafka-python/blob/71217575e65d5652508d80d0532314acd190fd85/confluent_kafka/avro/cached_schema_registry_client.py#L207

How to reproduce
================
Just looking at code...

Checklist
=========
Please provide the following information:

 - [ ] confluent-kafka-python and librdkafka version (`confluent_kafka.version()` and `confluent_kafka.libversion()`):
 - [ ] Apache Kafka broker version:
 - [ ] Client configuration: `{...}`
 - [ ] Operating system:
 - [ ] Provide client logs (with `'debug': '..'` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

### PR DESCRIPTION
**Pull request by Enether**
Fixes #372 

- Return `False` instead of `None` when unable to check compatibility
- Fix unassigned variable problem when request inside `get_compatibility()` returns a non-successful status code.

I tried to implement tests for `get_compatibility()` but I could not figure out what would be a proper way to implement `MockSchemaRegistryClient#get_compatibility()`, as I do not grasp the way the schema registry works and would rather not implement some hardcoded mock value like
```
if subject is None: return { 'compatibility': 'NONE', 'compatibilityLevel': -1}
else ...
```

### CODE REVIEW

**Comment by Enether** [comment]
[clabot:check]

**Comment by Enether** [comment]
The only test that failed in the Travis CI build is:
`tests/test_threads.py::test_thread_safety FAILED                         [ 53%]`
I assume this has nothing to do with my changes and maybe requires a re-run or re-configuration?

File: confluent_kafka/avro/cached_schema_registry_client.py [hunk 1]
```diff
@@ -259,16 +259,16 @@ def test_compatibility(self, subject, avro_schema, version='latest'):
...
             if code == 404:
-                log.error(("Subject or version not found:" + str(code)))
+                log.error("Subject or version not found:" + str(code))
                 return False
             elif code == 422:
-                log.error(("Invalid subject or schema:" + str(code)))
+                log.error("Invalid subject or schema:" + str(code))
                 return False
-            elif code >= 200 and code <= 299:
+            elif 200 <= code <= 299:
```
Comment (edenhill) [inline confluent_kafka/avro/cached_schema_registry_client.py]: Please avoid unrelated style changes.

File: confluent_kafka/avro/cached_schema_registry_client.py [hunk 2]
```diff
@@ -259,16 +259,16 @@ def test_compatibility(self, subject, avro_schema, version='latest'):
...
                 return result.get('is_compatible')
             else:
-                log.error("Unable to check the compatibility")
-                False
+                log.error("Unable to check the compatibility:" + str(code))
```
Comment (edenhill) [inline confluent_kafka/avro/cached_schema_registry_client.py]: Should be a space after ":"

File: confluent_kafka/avro/cached_schema_registry_client.py [hunk 3]
```diff
@@ -308,10 +308,9 @@ def get_compatibility(self, subject=None):
             url += '/' + subject
 
         result, code = self._send_request(url)
-        if code >= 200 and code <= 299:
+        if 200 <= code <= 299:
             compatibility = result.get('compatibility', None)
-
-        if not compatibility:
+        else:
```
Comment (edenhill) [inline confluent_kafka/avro/cached_schema_registry_client.py]: Don't think this is correct: result.get() may return None, in which case we should fall back to compatibilityLevel here.

File: confluent_kafka/avro/cached_schema_registry_client.py (hunk 2 above)
Comment (Enether) [reply confluent_kafka/avro/cached_schema_registry_client.py]: I also thought so but intentionally did not add it so it would be consistent with the two prints above

File: confluent_kafka/avro/cached_schema_registry_client.py (hunk 1 above)
Comment (Enether) [reply confluent_kafka/avro/cached_schema_registry_client.py]: 👍

File: confluent_kafka/avro/cached_schema_registry_client.py (hunk 3 above)
Comment (Enether) [reply confluent_kafka/avro/cached_schema_registry_client.py]: Yep, absolutely. Great catch, thanks

File: confluent_kafka/avro/cached_schema_registry_client.py [hunk 4]
```diff
@@ -311,7 +312,7 @@ def get_compatibility(self, subject=None):
         if code >= 200 and code <= 299:
             compatibility = result.get('compatibility', None)
 
-        if not compatibility:
+        if compatibility is None:
```
Comment (Enether) [inline confluent_kafka/avro/cached_schema_registry_client.py]: This changes the behavior if `result.get('compatibility', None)` returns an empty string, but I assume that's not possible

File: confluent_kafka/avro/cached_schema_registry_client.py [hunk 5]
```diff
@@ -311,7 +312,7 @@ def get_compatibility(self, subject=None):
         if code >= 200 and code <= 299:
             compatibility = result.get('compatibility', None)
```
Comment (edenhill) [inline confluent_kafka/avro/cached_schema_registry_client.py]: Looking at the REST API spec it seems that this value is never returned but `is_compatible`:
https://docs.confluent.io/current/schema-registry/docs/api.html#id1

There is also no mention of `compatibilityLevel` being returned.

So either the documentation is wrong or this function must never have worked.

Comment (Enether) [reply confluent_kafka/avro/cached_schema_registry_client.py]: The REST API was updated. It did return `compatibility` in version `3.0` - https://docs.confluent.io/3.0.1/schema-registry/docs/api.html#config
And yeah, `compatibilityLevel` is not mentioned anywehere, even in version `2.0`

Comment (Enether) [reply confluent_kafka/avro/cached_schema_registry_client.py]: Wow haha, nope. We're looking at the wrong API endpoint.
This calls GET /config
https://docs.confluent.io/current/schema-registry/docs/api.html#get--config
Which does return `compatibility`. Still no mention of `compatibilityLevel` though

Comment (Enether) [reply confluent_kafka/avro/cached_schema_registry_client.py]: The question then becomes - what should we return when the response is a non  2xx status code - 'NONE' ?

File: confluent_kafka/avro/cached_schema_registry_client.py [hunk 6]
```diff
@@ -301,17 +301,11 @@ def get_compatibility(self, subject=None):
         Get the current compatibility level for a subject.  Result will be one of:
 
         @:param: subject: subject name
-        @:return: 'NONE','FULL','FORWARD', or 'BACKWARD'
+        @:return: True if compatible, False if not compatible
```
Comment (edenhill) [inline confluent_kafka/avro/cached_schema_registry_client.py]: Let's follow what the Java code does:
https://github.com/confluentinc/schema-registry/blob/1443a7f8e251c93987099b26b0917bcebd4deca5/client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java#L315

 /config returns a JSON  result containing "compatibility". getCompatibilityLevel() wraps the result and returns that same "compatibility" (and not "compatibilityLevel"), unfortunate naming perhaps.

And I propose raising an exception if code != 2xx or if the returned value is not one of the four strings.

File: confluent_kafka/avro/cached_schema_registry_client.py (hunk 5 above)
Comment (Enether) [reply confluent_kafka/avro/cached_schema_registry_client.py]: This method calls '/config':
https://github.com/confluentinc/confluent-kafka-python/blob/66a181459731515dc0bb86a9de69dcc51278797c/confluent_kafka/avro/cached_schema_registry_client.py#L300
https://github.com/confluentinc/confluent-kafka-python/blob/master/confluent_kafka/avro/cached_schema_registry_client.py#L306

**Comment by edenhill** [comment]
Thank you for this! :)
